api/cluster/controller.go

```diff
@@ -257,6 +257,11 @@ func (c *controller) Deploy(ctx context.Context, modelService *models.Service) (
 			log.Errorf("unable to create pdb: %v", err)
 			return nil, errors.Wrapf(err, fmt.Sprintf("%v", ErrUnableToCreatePDB))
 		}
+
+		unusedPdbs := getUnusedPodDisruptionBudgets(modelService, pdbs)
+		if err := c.deletePodDisruptionBudgets(ctx, unusedPdbs); err != nil {
+			log.Warnf("unable to delete unused pdb: %v", err)
```

Comment:
If the name of the PDB that want to be deleted is not exist, this will not return an error. But if for any other reason this call throw an error, I only give a log warning because the previous version PDB will not affect the newly deployed model version anyway (see **notes** in PR description for detail)

Reply:
can you add pdb name in the warning log?

Reply:
I’m a bit unsure what’s the best way to do this. The `deletePodDisruptionBudgets` is used to delete multiple PDBs, will it be odd to return only one name `(name string, err error)` for logging purpose, or is that fine? 

Should I iterate and delete the pdb like this? But I think it will be redundant with `deletePodDisruptionBudgets` function
```go
unusedPdbs := getUnusedPodDisruptionBudgets(modelService, pdbs)
for _, pdb := range unusedPdbs {
 if err := c.deletePodDisruptionBudget(ctx, pdb); err != nil {
  log.Warnf("unable to delete unused pdb %s: %v", pdb.Name, err)
 }
}
```
 Do you have any advice?

Reply:
I think merlin model and version should sufficient

Reply:
Actually I've got a completely different take on this - before calling `deployPodDisruptionBudgets`, can we simply delete all existing PDBs using `deletePodDisruptionBudgets`? That way we don't need to deal with additional logic in `getUnusedPodDisruptionBudgets`? I'm referring to something like this:
```go
		existingPDBs := getExistingPodDisruptionBudgets(modelService)
		if err := c.deletePodDisruptionBudgets(ctx, existingPDBs); err != nil {
			// do something
		}
		// Create new pdbs
		pdbs := generatePDBSpecs(modelService, c.deploymentConfig.PodDisruptionBudget)
		if err := c.deployPodDisruptionBudgets(ctx, pdbs); err != nil {
			log.Errorf("unable to create pdb: %v", err)
			return nil, errors.Wrapf(err, fmt.Sprintf("%v", ErrUnableToCreatePDB))
		}
```

I see two ways to get the existing PDBs, one is to use the Kube API to retrieve a list of actual PDBs for that deployment (potentially using a label selector like this: https://github.com/caraml-dev/merlin/blob/24d79eb81a3a9e373f3c12bcf134a982231c03c0/api/cluster/pdb.go#L30), OR to simply generate a list of potential PDB names (only 2 combinations really, model only or model+transformer) since we already have the naming convention for PDBs, and then attempt to delete all of them. In the second case, we can choose to not return any error (and just log them if necessary).

Reply:
Oh and I don't want to ruin the party, but we might have to consider porting these changes to Turing too 👀

Reply:
Okay, got it!

Reply:
Oh, I didn’t see your @deadlycoconuts  comment when sending my latest comment lol.

When we run the delete command first, it could yield potential extra operations. E.g. previously there’s both the predictor & transformer PDB, we will then delete that two PDBs, and recreate two PDBs. While if we create it first, we will only need to patch the PDBs (because it has the same name), and delete nothing. There’s also a lookup operation if we want to use the label selector. My question is, considering there’s a lot of things happening when deploying a model already, is the potential operations is fine? If yes, I think we can move with your suggestion bcs the code looks more intuitive. 
> In either case, we would need need to update the implementation of this method here (called when undeploying Merlin models) too, so that we aren't deleting only PDBs that would've been created by generatePDBSpecs:

Does this mean if there’s a PDB not created by `generatePDBSpecs`, we will delete it to? 

And, what’s going on with Turing? 😅

Reply:
Haha no worries lol, I'm not constantly following the comments in this PR either 😂

Yeah what you said it's true, we'll definitely be making more calls to the Kube API server and not take advantage of the 'patch' operation in k8s by first deleting the existing PDBs before recreating them (if any). To be honest I think any additional operations should be fine, if you were to perform a list operation with the label selector to get the existing PDBs to delete them, that'll only be 3 more API calls at most, which is fine (adding/removing PDBs is also an almost instantaneous operation unlike adding/removing pods). 

> Does this mean if there’s a PDB not created by generatePDBSpecs, we will delete it to?

Hmm I thought over about this again, okay it's fine to keep it as it is; with your fix, all PDBs should be correct (there shouldn't be undeleted PDBs) with respect to the current deployment's resource requests, so using `generatePDBSpecs` would generate the correct set of PDBs to delete. (I was thinking of the possibility of just deleting all PDBs found for that particular model version but I guess that's unnecessary once we fix the redeployment workflow).

Oh and regarding the Turing comment, my bad, I checked how PDBs were created/deleted again and Turing actually doesn't have this bug because of the way how Turing tracks every single modification to a router as a version change (so its PDBs are always removed and recreated with a different name), unlike in Merlin where you can redeploy the same model version with a different set of deployment configs.

Reply:
After a bit of discussion with @tiopramayudi, I think it will be safer to delete the PDB once the [new inference service is ready](https://github.com/caraml-dev/merlin/blob/24d79eb81a3a9e373f3c12bcf134a982231c03c0/api/cluster/controller.go#L262). 

Suppose that for any reason the new inference service is not ready (maybe bcs of timeout), Merlin will not going to serve the new model revision, it will go back to serve the current model revision instead. If we delete the PDB first, then the PDB that is used for the current revision is now already gone.

But, there is another issue with this case. If the new PDB is deployed but `waitInferenceServiceReady` got a timeout, the PDB now will not match the current pod's label anyway (because PDB has the same name, and the label selector has been replaced to the fail inference service's revisionID ([ref1](https://github.com/caraml-dev/merlin/blob/24d79eb81a3a9e373f3c12bcf134a982231c03c0/api/cluster/pdb.go#L30) & [ref2](https://github.com/caraml-dev/merlin/blob/24d79eb81a3a9e373f3c12bcf134a982231c03c0/api/cluster/pdb.go#L58))). 

Here’s what I’m thinking of. First, the name of the PDB will include the revision number (again):
```
{modelName}-{modelVersion}-{revisionID}-{componentType}-pdb
```

So, this new PDB will not patch the config of the current PDB. Then, once the inference service is up and everything is okay, the unused PDB will be deleted, can put it at the very end of the function, maybe right before [this](https://github.com/caraml-dev/merlin/blob/24d79eb81a3a9e373f3c12bcf134a982231c03c0/api/cluster/controller.go#L299). 

```go
unusedPDBs := getUnusedPodDisruptionBudgets(modelService)
if err := c.deletePodDisruptionBudgets(ctx, existingPDBs); err != nil {
	// log if error
}

```
I’m still a bit unsure on how to get the unused PDBs efficiently though because we need to filter it out based on model name & version, but the PDB label doesn’t have a version field. Maybe can use label selector where `app = {model name} AND inferenceservice != {inference service name}`, then filter out the name which doesn’t match the version number with regex.

Reply:
Got it I agree that it's safer to delete the PDB only after the new inference is ready, and that we can delete the PDBs at the very end of the function. What about doing it [here](https://github.com/caraml-dev/merlin/blob/24d79eb81a3a9e373f3c12bcf134a982231c03c0/api/cluster/controller.go#L293), just after deleting the old inference service? 

I also like the idea of reintroducing the revision number back into the PDB names again so that makes it much easier for us to retrieve and delete the correct PDBs. 

> I’m still a bit unsure on how to get the unused PDBs efficiently though because we need to filter it out based on model name & version, but the PDB label doesn’t have a version field. Maybe can use label selector where app = {model name} AND inferenceservice != {inference service name}, then filter out the name which doesn’t match the version number with regex.

Hmm can we just attempt to perform 2 deletion operations? We can construct the names of the (existing) PDBs entirely using whatever's contained within the `modelService` object - it contains `CurrentIsvcName` (which is the existing isvc's name), which is equivalent to `{modelName}-{modelVersion}-{revisionID}` already. All we need to do is append `-{componentType}-pdb` and then attempt to delete both of them.

Reply:
Hi I’ve updated the code for this case 😁

I said in the comment that the PDB’s `label` and `matchLabels.Selector` are the same, but as I’ve mentioned in our meeting, I think it will be simpler to modify that. So, Merlin will add `model-version-id` labels to PDB, but the `selector` will follow current behavior. With this approach, the `getUnusedPodDisruptionBudgets` will get all unused PDB based on model name and version, instead of just the previous latest version PDB (in case of deletion failure, it can be picked up another time user redeploy the endpoint).

The name of the PDB will still use a revisionID so `{modelName}-{modelVersion}-{revisionID}-{componentType}-pdb` and I put the deletion code after the line you mentioned.

And I’ll get back again after finding out how to remove all the other PDB from previous/current Merlin’s version that is still running in our cluster